src/entity/Living.php

```diff
@@ -293,10 +293,26 @@ public function saveNBT() : CompoundTag{
 		return $nbt;
 	}
 
+	/**
+	 * Returns whether this entity has not block line of sight to another entity.
+	 *
+	 * @throws \InvalidArgumentException if the entities are not in the same world
+	 */
 	public function hasLineOfSight(Entity $entity) : bool{
-		//TODO: head height
+		$world = $this->getWorld();
+		if ($world !== $entity->getWorld()) {
+			throw new \InvalidArgumentException("Target entity is not in the same world");
+		}
+		$start = $this->getEyePos();
+		$end = $entity->getEyePos();
+		if ($end->subtractVector($start)->normalize()->lengthSquared() > 0) {
+			foreach(VoxelRayTrace::betweenPoints($start, $end) as $vector3){
+				if(!$world->getBlockAt((int) floor($vector3->x), (int) floor($vector3->y), (int) floor($vector3->z))->isTransparent()){
```

Comment:
floor() is not necessary here, VoxelRayTrace will always return whole number vectors.

Reply:
I doubt that isTransparent is suitable here, because objects like chests are technically transparent, but aren't actually see-through. isTransparent is a property relevant only for light propagation.